### ISSUE TITLE
Faked service is not being recognized in classes outside of test that load the dependency

### ISSUE THREAD

**Issue by mattmallon** [poppinss/adonis-fold#13]
## Behavior
Hi, I'm attempting to mock a service in an Adonis test by using ioc.fake, as documented in the Adonis testing documentation, but the behavior I'm seeing is that the class is only mocked within my test, and it is not mocked within the other classes under test that rely on the fake for a dependency. I am importing the class as I typically would with a `use()` statement in the other classes, rather than passing in the mock as an argument to a class constructor. The documentation made it seem like that was the correct use case...? Am I supposed to instead create a mock within my test and then manually pass it in as a dependency to any class that relies on it?

I thought that maybe the issue was that I had to mock the service before I include other classes through `use()` that rely on the mocked service as a dependency, but when I tried that, I saw the same behavior. 

I've included code below that hopefully will illustrate my point if this doesn't quite make sense yet. I added a comment that says `//ISSUE` that traces the path through to where I'm seeing the actual class method being called rather than the fake class method. I also added console log statements for debugging; when I run `adonis test` I see `in real getCourse` printed from the actual function rather than `in fake getCourse` which should be printed from the fake function call. Thank you!

## Package version
Fold: 4.0.9
Adonis: 5.0.8
Vow: 1.0.17

## Node.js and npm version
Node: 10.8.0
Npm: 6.2.0

## Sample Code (to reproduce the issue)

*example.spec.js:*
```
'use strict'

const { ioc } = use('@adonisjs/fold');
const { afterEach, before, beforeEach, test, trait } = use('Test/Suite')('Assignment due notification');

const Course = use('App/Models/Course');

let course;

//wraps all DB transactions and rolls them back after test completes
trait('DatabaseTransactions');

beforeEach(async () => {

    ioc.fake('App/Classes/Canvas/CanvasApi', () => {
        return {
            async getCourse() {
                console.log('in fake getCourse');
                return {
                    'name': 'Fake test course',
                    'term': {name: '2018'},
                    'time_zone': 'America/New_York',
                    'workflow_state': 'available'
                };
            }
        }
    });

    //DEBUGGING HERE: the mock shows up correctly within this test function,
    //but it does not show up correctly in any other class that uses it.
    //I have tried mocking the class here before loading in any dependencies,
    //and that did not change the behavior (accounted as well for variable hoisting)
    console.log('test in beforeEach: ');
    const canvasApi = use('App/Classes/Canvas/CanvasApi');
    console.log(canvasApi);

    const courseId = '000';
    const contextId = 'fake-lti-context-id';
    course = new Course();
    //ISSUE in the function below
    await course.create(courseId, contextId);

    ioc.restore('App/Classes/Canvas/CanvasApi'); //restore so class can be re-mocked within individual tests
});

test('it should not send notifications if no assignments are upcoming', async ({ assert }) => {
    //tests will eventually go here
});
```

*App/Models/Course.js:*
```
'use strict'

const Model = use('Model');
const CanvasApi = use('App/Classes/Canvas/CanvasApi');
const Config = use('Config');

class Course extends Model {
    courseNotificationSettings() {
        return this.hasMany('App/Models/CourseNotificationSetting');
    }

    async create(courseId, contextId) {
        //ISSUE in the function below
        const canvasCourseData = await this.getCanvasCourseData(courseId, {includeTerm: true});

        this.context_id = contextId;
        this.canvas_course_id = courseId;
        this.canvas_course_name = canvasCourseData.name;
        this.canvas_course_term = canvasCourseData.term.name;
        this.canvas_course_time_zone = canvasCourseData.time_zone;
        this.canvas_course_workflow_state = canvasCourseData.workflow_state;
        this.enabled = 0;

        await this.save();

        return this;
    }

    async getCanvasCourseData(courseId, options) {
        const canvasApi = new CanvasApi();
        //ISSUE in the function below
        return await canvasApi.getCourse(courseId, options);
    }
}

module.exports = Course;

```

*App/Classes/Canvas/CanvasApi.js:*
```
'use strict';

const Env = use('Env');
const axios = require('axios');
const moment = require('moment');
const parseLinkHeader = require('parse-link-header');

class CanvasApi {
    constructor() {
        this.apiToken = Env.get('CANVAS_API_TOKEN');
        this.canvasDomain = Env.get('CANVAS_API_DOMAIN');
    }

    async getCourse(courseId, options = {}) {
        //ISSUE HERE -- should not be called
        console.log('in real getCourse');
        const urlPrefix = `/courses/${courseId}`;
        let queryParams = [];
        if (options.includeTerm) {
            queryParams.push({ 'key': 'include[]', 'value': 'term' });
        }

        const results = await this.canvasGetRequest(urlPrefix, queryParams);
        return results;
    }

    async canvasGetRequest(urlPrefix, queryParams = []) {
        let url = `${this.canvasDomain}${urlPrefix}?per_page=100`;
        for (let queryParam of queryParams) {
            url += `&${encodeURIComponent(queryParam.key)}=${encodeURIComponent(queryParam.value)}`;
        }
        const headers = this.getAuthorizationHeader();
        let response = null;
        let returnedData = [];
        let requestsRemaining = true;
        let nextPaginationLink = false;
        let requestsMade = 0;

        while (requestsRemaining) {
            try {
                response = await axios.get(url, { headers: headers });

                //if an array, we may need to combine paginated results;
                //otherwise, just return the single or empty resource as is
                if (Array.isArray(response.data)) {
                    returnedData = returnedData.concat(response.data);
                }
                else {
                    returnedData = response.data;
                }

                requestsMade++;
                nextPaginationLink = this.getNextPaginationLink(response);

                if (!nextPaginationLink) {
                    requestsRemaining = false;
                }
                else {
                    url = nextPaginationLink.url;
                }

                //fail-safe here so we don't accidentally end up in an infinite loop or
                //land upon some Canvas resource that is too huge for us to handle
                if (requestsMade > 50) {
                    throw new Error('Exiting Canvas API, too many requests made to url: ' + url);
                }
            }
            catch (error) {
                requestsRemaining = false;
                throw new Error(this.getErrorMessage(error));
            }
        }

        if (returnedData.length < 2) {}

        return returnedData;
    }

    getNextPaginationLink(response) {
        const paginationLink = response.headers.link;
        let nextPaginationLink = false;

        if (paginationLink) {
            const parsedPaginationLink = parseLinkHeader(paginationLink);
            nextPaginationLink = parsedPaginationLink.next;
        }

        return nextPaginationLink;
    }

    getAuthorizationHeader() {
        return { "Authorization": `Bearer ${this.apiToken}` };
    }

    getErrorMessage(error) {
        let message = '';

        for (let errorMessage of error.response.data.errors) {
            message += (errorMessage.message + ' ');
        }

        return message;
    }
}

module.exports = CanvasApi;
```

**Comment by tavsta** [comment]
Is there any solution?

**Comment by thetutlage** [comment]
Fixed in `v6.0.0`

**Comment by mattmallon** [comment]
@thetutlage Hi Harminder, thank you for the update. When I install a version of Vow >= 6.0.0 in a project, I get the following error from Ignitor when I try to start the server:

```
TypeError: Cannot read property 'appNamespace' of undefined
 
at /home/node/app/server.js(anonymous):24
19
20  const { Ignitor } = require('@adonisjs/ignitor')
21
22  new Ignitor(require('@adonisjs/fold'))
23    .appRoot(__dirname)
24    .fireHttpServer()
25    .catch(console.error)
26
 
1 Ignitor._setupResolver
  /home/node/app/node_modules/@adonisjs/ignitor/src/Ignitor/index.js:172
 
2 Object.keys.forEach
  /home/node/app/node_modules/@adonisjs/ignitor/src/Ignitor/index.js:210
 
3 Ignitor._registerAutoloadedDirectories
  /home/node/app/node_modules/@adonisjs/ignitor/src/Ignitor/index.js:207
 
4 Ignitor.fire
  /home/node/app/node_modules/@adonisjs/ignitor/src/Ignitor/index.js:725
 
5 Ignitor.fireHttpServer
  /home/node/app/node_modules/@adonisjs/ignitor/src/Ignitor/index.js:791
```

Here are the steps to reproduce:

1. `adonis new foldtest`
2. `cd foldtest`
3. `npm install @adonisjs/fold@6.0.3`
4. `adonis serve --dev`

When running tests, I get errors about being unable to find installed providers. For instance, after running `adonis install @adonisjs/vow` and then running `adonis test`:

```
Error: Cannot find module '@adonisjs/vow/providers/VowProvider'


1 requireStack
  /home/node/app/node_modules/require-stack/src/index.js:44

2 _.uniq.map
  /home/node/app/node_modules/@adonisjs/fold/src/Registrar/index.js:104

3 arrayMap
  /home/node/app/node_modules/lodash/lodash.js:639

4 Function.map
  /home/node/app/node_modules/lodash/lodash.js:9556

5 interceptor
  /home/node/app/node_modules/lodash/lodash.js:16993

6 thru
  /home/node/app/node_modules/lodash/lodash.js:8797

7 anonymous
  /home/node/app/node_modules/lodash/lodash.js:4374

8 arrayReduce
  /home/node/app/node_modules/lodash/lodash.js:683

9 baseWrapperValue
  /home/node/app/node_modules/lodash/lodash.js:4373

10 LazyWrapper.lazyValue [as value]
  /home/node/app/node_modules/lodash/lodash.js:1859

11 baseWrapperValue
  /home/node/app/node_modules/lodash/lodash.js:4371

12 LodashWrapper.wrapperValue
  /home/node/app/node_modules/lodash/lodash.js:9052
```

Are the newer versions of Vow not intended to be used quite yet? I noticed that the latest official version on npm was 4.0.9 rather than the actual latest, 6.0.3. Unfortunately these issues with Vow are preventing me from writing unit tests at the moment. Thank you for your help!
